### PR TITLE
Integrate CST: Work in progress

### PR DESCRIPTION
**Pull request by mdevils**
Continuing CST **Work In Progress** pull request. Want to keep you informed about the current integration I'm working on.

In this pull request:

* Use CST instead of AST with token hacks (https://github.com/mdevils/cst)
* Simplify JsFile and rules.
* Improve assertions in tests using chai and chai extensions (it is very hard to make huge changes with the current error messages).

Feel free to discuss changes. Any help is very appreciated.

I try to keep everything rebased.

/cc @markelog @hzoo @mrjoelkemp @mikesherov @zxqfox

### CODE REVIEW

File: lib/js-file.js [hunk 1]
```diff
@@ -110,24 +123,25 @@ JsFile.prototype = {
...
-                return true;
-            }
-
-            if (region.rule === ruleName || region.rule === '*') {
-                enabled = region.enabled;
-            }
-        }, this);
-
-        return enabled;
+    isEnabledRule: function(ruleName, element) {
```
Comment (markelog) [inline lib/js-file.js]: Are you planning to rework these methods as well? Seems unrelated to proposed changes?

File: lib/js-file.js [hunk 2]
```diff
@@ -415,11 +408,11 @@ JsFile.prototype = {
      */
     getNodesByType: function(type) {
         if (typeof type === 'string') {
-            return this._index[type] || [];
+            return this._program.selectNodesByType(type);
         } else {
             var result = [];
             for (var i = 0, l = type.length; i < l; i++) {
-                var nodes = this._index[type[i]];
+                var nodes = this._program.selectNodesByType(type[i]);
```
Comment (markelog) [inline lib/js-file.js]: Looks much better, although it seems there is no need for the `for` anymore, some of the `Array#<methods>` or `lodash` methods should do the job?

File: lib/js-file.js [hunk 3]
```diff
@@ -447,17 +440,18 @@ JsFile.prototype = {
...
-        var typeIndex = {};
-        types.forEach(function(type) {
-            typeIndex[type] = true;
-        });
-
-        this.getTokens().forEach(function(token, index, tokens) {
-            if (typeIndex[token.type]) {
-                cb(token, index, tokens);
+        var tokens;
+        if (Array.isArray(type)) {
```
Comment (markelog) [inline lib/js-file.js]: It seems argument check should look like the same way as in `getNodesByType` method

File: lib/js-file.js [hunk 4]
```diff
@@ -581,7 +575,15 @@ JsFile.prototype = {
      * Returns comment token list, built using esprima.
      */
     getComments: function() {
-        return this._tree.comments;
+        var comments = [];
+        var token = this._program.firstToken;
```
Comment (markelog) [inline lib/js-file.js]: Maybe method like that should be included in cst?

File: lib/js-file.js [hunk 5]
```diff
@@ -670,152 +649,6 @@ JsFile.prototype = {
...
-    },
-
-    /**
-     * Applies whitespace information to the token list.
-     *
-     * @param {Object[]} tokens
-     * @param {String} source
-     * @private
-     */
-    _applyWhitespaceData: function(tokens, source) {
```
Comment (markelog) [inline lib/js-file.js]: Was this method being replaced?

File: test/chai-extensions.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
+var chai = require('chai');
+var utils = chai.utils;
+
+chai.use(function(chai, utils) {
+    /**
+     * Error assertion for `Errors` instances.
+     */
+    chai.Assertion.addChainableMethod('error', function(message) {
```
Comment (markelog) [inline test/chai-extensions.js]: Would document those extensions at the wiki page, otherwise this could get confusing

**Comment by markelog** [comment]
Since changes are pretty big and will get bigger, i think we need to make this green for the travis - add all those `skip`'s or disable testing for the yet unmodified rules. And land it to `3.0` otherwise diff would became pretty big so it would be impossible to review properly.

File: lib/js-file.js (hunk 2 above)
Comment (mdevils) [inline lib/js-file.js]: This is a very frequently called function so I decided to have a fast `for`.

File: lib/js-file.js (hunk 4 above)
Comment (mdevils) [inline lib/js-file.js]: Yep, probably yes.

File: lib/js-file.js (hunk 5 above)
Comment (mdevils) [inline lib/js-file.js]: Yes, whitespace are tokens now, so no `whitespaceBefore` anymore.

File: test/chai-extensions.js (hunk 6 above)
Comment (mdevils) [inline test/chai-extensions.js]: +1

**Comment by mdevils** [comment]
@markelog OK, I will make it green soon.

File: lib/js-file.js (hunk 1 above)
Comment (mdevils) [inline lib/js-file.js]: I have to rewrite it somehow, since comments don't have fixed positions anymore.

**Comment by hzoo** [comment]
Babel moved comment attachment to the parser instead of estraverse to allow for comments in empty blocks - check out the recent babel-eslint changes
